### PR TITLE
WT-2031 Log slot revamp

### PR DESCRIPTION
**Pull request by sueloverso**
This is not ready to merge.  Some basic tests are working (ex_log and log-append.wtperf), but some examples and unit tests fail or hang still.  Creating a pull request in case you want to look at the changes so far.

### CODE REVIEW

File: src/include/log.h [hunk 1]
```diff
@@ -48,48 +47,77 @@
...
  */
-#define	WT_LOG_SLOT_DONE	0
-#define	WT_LOG_SLOT_FREE	1
-#define	WT_LOG_SLOT_PENDING	2
-#define	WT_LOG_SLOT_WRITTEN	3
-#define	WT_LOG_SLOT_READY	4
+#define	WT_LOG_SLOT_BITS	2
+#define	WT_LOG_SLOT_MAXBITS	(32 - WT_LOG_SLOT_BITS)
+#define	WT_LOG_SLOT_CLOSE	0x4000000000000000	/* Force slot close */
+#define	WT_LOG_SLOT_RESERVED	0x8000000000000000	/* Reserved states */
```
Comment (michaelcahill) [inline src/include/log.h]: Do these constants need a `ULL` suffix?

File: src/log/log.c [hunk 2]
```diff
@@ -697,6 +676,139 @@ err:	__wt_scr_free(session, &from_path);
...
+	 */
+	log->alloc_lsn.file = log->fileid;
+	log->alloc_lsn.offset = WT_LOG_FIRST_RECORD;
+	end_lsn = log->alloc_lsn;
+
+	/*
+	 * If we're called from connection creation code, we need to update
+	 * the LSNs since we're the only write in progress.
+	 */
+	if (conn_create) {
```
Comment (michaelcahill) [inline src/log/log.c]: The naming here puzzled me -- if I'm understanding correctly, `conn_create` indicates that we're creating a log file as part of opening a database (i.e., inside `wiredtiger_open`).  I guess we're creating a connection handle, but `conn_open` would make more sense to me.

**Comment by michaelcahill** [comment]
One general comment: I tried to build on OS X with all warnings turned on and saw quite a few `-Wsign-conversion` and a few `-Wshorten-64-to-32` warnings.  Can you please take a look?

File: test/suite/test_txn04.py [hunk 3]
```diff
@@ -113,6 +113,7 @@ def hot_backup(self, backup_uri, committed):
         # If we are backing up a target, assume the directory exists.
         # We just use the wt backup command.
         # A future test extension could also use a cursor.
+        self.runWt(['-h', self.home, 'printlog'], outfilename='printlog.out')
```
Comment (michaelcahill) [inline test/suite/test_txn04.py]: Is this just debugging?

File: src/log/log_slot.c [hunk 4]
```diff
@@ -106,211 +305,115 @@ __wt_log_slot_join(WT_SESSION_IMPL *session, uint64_t mysize,
...
+	}
 	/*
-	 * Add my size into the state.  When it reaches WT_LOG_SLOT_DONE
-	 * all participatory threads have completed copying their piece.
+	 * Add my size into the state and return the new size.
 	 */
-	return (__wt_atomic_addiv64(&slot->slot_state, (int64_t)size));
+	my_size = WT_LOG_SLOT_JOIN_REL((uint64_t)0, size, 0);
+	newsize = __wt_atomic_addiv64(&slot->slot_state, my_size);
+	return (newsize);
```
Comment (michaelcahill) [inline src/log/log_slot.c]: Looks like the local variable `newsize` only exists for debugging?

File: src/include/extern.h [hunk 5]
```diff
@@ -354,13 +356,14 @@ extern int __wt_logop_row_truncate_pack( WT_SESSION_IMPL *session, WT_ITEM *logr
 extern int __wt_logop_row_truncate_unpack( WT_SESSION_IMPL *session, const uint8_t **pp, const uint8_t *end, uint32_t *fileidp, WT_ITEM *startp, WT_ITEM *stopp, uint32_t *modep);
 extern int __wt_logop_row_truncate_print( WT_SESSION_IMPL *session, const uint8_t **pp, const uint8_t *end, FILE *out);
 extern int __wt_txn_op_printlog( WT_SESSION_IMPL *session, const uint8_t **pp, const uint8_t *end, FILE *out);
+extern void __wt_log_slot_activate(WT_SESSION_IMPL *session, WT_LOGSLOT *slot);
+extern int __wt_log_slot_close(WT_SESSION_IMPL *session, WT_LOGSLOT *slot, int *releasep);
+extern int __wt_log_slot_switch(WT_SESSION_IMPL *session, WT_LOGSLOT *slot);
+extern int __wt_log_slot_new(WT_SESSION_IMPL *session);
```
Comment (michaelcahill) [inline src/include/extern.h]: Somewhat peripheral, but can you check whether all of these need to be exported from `log.c`?  As far as I can see, most of them could be `static`.

File: src/include/log.h [hunk 6]
```diff
@@ -48,48 +47,77 @@
...
+
+#define	WT_LOG_SLOT_MASK_OFF	0x3fffffffffffffff
+#define	WT_LOG_SLOT_MASK_ON	~(WT_LOG_SLOT_MASK_OFF)
+
+#define	WT_LOG_SLOT_MAXIMUM		(uint32_t)			\
+    ((2 ^ WT_LOG_SLOT_MAXBITS) - WT_LOG_SLOT_BUF_SIZE)
```
Comment (michaelcahill) [inline src/include/log.h]: What is this number supposed to be?  I don't think you want the XOR in `2 ^ 30 == 28`...

File: src/include/log.h [hunk 7]
```diff
@@ -48,48 +47,77 @@
...
+ * The slot state is divided into two 32 bit sizes.  One half is the
+ * amount joined and the other is the amount released.  Since we use
+ * a few special states, reserve the top few bits for state.
+ */
+#define	WT_LOG_SLOT_BUF_SIZE		(256 * 1024)
+#define	WT_LOG_SLOT_BUF_MAX		((uint32_t)log->slot_buf_size / 2)
+
+/*
+ * The high bit is reserved for the special states.
+ */
```
Comment (michaelcahill) [inline src/include/log.h]: Could you please expand the comments through this section?

File: src/log/log.c [hunk 8]
```diff
@@ -1105,7 +1245,17 @@ __log_release(WT_SESSION_IMPL *session, WT_LOGSLOT *slot, int *freep)
...
+#if 0
+		/*
+		 * If it is not equal, it better not be larger than our
+		 * release LSN.  Check <= because it could have changed
+		 * to now be equal.
+		 */
+		WT_ASSERT(session,
+		    __wt_log_cmp(&log->write_lsn,
+		    &slot->slot_release_lsn) <= 0);
+#endif
```
Comment (michaelcahill) [inline src/log/log.c]: Will you remove this?

**Comment by michaelcahill** [comment]
@sueloverso, I've read through this today -- it is making sense to me, but there are some details I haven't completely understood.  I've asked a few questions, and I mentioned the compiler warnings.  If you can deal with those, I'm happy for this to be merged and I can study it again afterwards.

File: src/include/extern.h (hunk 5 above)
Comment (sueloverso) [inline src/include/extern.h]: I would prefer to keep functions that manipulate slots in `log_slot.c` even if their only callers are in `log.c` (and that only applies to `__wt_log_slot_switch`,  the others are called from both).

File: src/include/log.h (hunk 6 above)
Comment (sueloverso) [inline src/include/log.h]: Fixed.

File: src/include/log.h (hunk 7 above)
Comment (sueloverso) [inline src/include/log.h]: Done.  Read through after I push and see if it gives you the information you want.

**Comment by sueloverso** [comment]
I have taken care of most of the warnings.  My Mac's gcc (4.2.1) doesn't have either warning setting.  The AWS box has `-Wsign-conversion` but not `-Wshorten-64-to-32` (I checked both that I use).   I will merge once pull-request-tester succeeds.

**Comment by sueloverso** [comment]
@michaelcahill Please review the changes in 228e315.  I found this by inspection when looking at the log compares yesterday.  A sanity check would be good.

**Comment by sueloverso** [comment]
Ok.  I will merge once the tester completes (for real this time)!

**Pull request merged**